executor/runtime/docker/docker.go

```diff
@@ -558,8 +558,36 @@ func sleepWithCtx(parentCtx context.Context, d time.Duration) error {
 	}
 }
 
-func (r *DockerRuntime) dockerPull(ctx context.Context, c *runtimeTypes.Container) error {
-	return pullWithRetries(ctx, r.metrics, r.client, c, doDockerPull)
+func imageExists(ctx context.Context, client *docker.Client, ref string) (bool, error) {
+	resp, _, err := client.ImageInspectWithRaw(ctx, ref)
+	if err != nil {
+		if docker.IsErrImageNotFound(err) {
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	log.Debugf("Image '%s' exists. Response: %+v", ref, resp)
+	return true, nil
+}
+
+// DockerPull returns true if the image was cached, and we didn't need to pull
+func (r *DockerRuntime) DockerPull(ctx context.Context, c *runtimeTypes.Container) (bool, error) {
+	imgName := c.QualifiedImageName()
+	imgExists, err := imageExists(ctx, r.client, imgName)
```

Comment:
We should only do this for digests. Right now, the behaviour with labels is we always try to re-pull the image in case somewhat mutated it.